File: Java e python/Técnicas/Nodes/InorderSemPilha(Morris_Traversal).java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; }
}

public class TechniquesTree {
    /**
     * Inorder Morris Traversal sem pilha nem recursão.
     */
    public List<Integer> inorderMorris(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        TreeNode curr = root;
        while (curr != null) {
            if (curr.left == null) {
                res.add(curr.val);
                curr = curr.right;
            } else {
                // predecessor
                TreeNode pred = curr.left;
                while (pred.right != null && pred.right != curr) {
                    pred = pred.right;
                }
                if (pred.right == null) {
                    // cria thread
                    pred.right = curr;
                    curr = curr.left;
                } else {
                    // destrói thread
                    pred.right = null;
                    res.add(curr.val);
                    curr = curr.right;
                }
            }
        }
        return res;
    }
}

File: Java e python/Técnicas/Nodes/InorderSemPilha(Morris_Traversal).py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorder_morris(root: TreeNode) -> list[int]:
    """
    - Constrói links temporários (threads) para voltar ao pai sem pilha.
    - Tempo: O(n), Espaço: O(1) extra.
    """
    res = []
    curr = root
    while curr:
        if curr.left is None:
            # visita e vai para direita
            res.append(curr.val)
            curr = curr.right
        else:
            # acha o predecessor (máx da subárvore esquerda)
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if pred.right is None:
                # faz o thread para voltar
                pred.right = curr
                curr = curr.left
            else:
                # thread existente: remove, visita, vai para direita
                pred.right = None
                res.append(curr.val)
                curr = curr.right
    return res

# Exemplo:
#    1
#     \
#      2
#     /
#    3
root = TreeNode(1, None, TreeNode(2, TreeNode(3), None))
print(inorder_morris(root))  # [1,3,2]
